File: cpp/src/grid_context.cpp
#include "grid_context.hpp"
#include "longbridge.h"
#include "convert.hpp"
#include "utils.hpp"
#include <algorithm>
#include <iterator>

namespace longbridge {
namespace grid {

using longbridge::convert::convert;
using longbridge::convert::convert_grid_trade_rule;

GridContext::GridContext()
  : ctx_(nullptr)
{
}

GridContext::GridContext(const lb_grid_context_t* ctx)
{
  ctx_ = ctx;
  if (ctx_)
    lb_grid_context_retain(ctx_);
}

GridContext::GridContext(const GridContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_)
    lb_grid_context_retain(ctx_);
}

GridContext::GridContext(GridContext&& ctx)
{
  ctx_ = ctx.ctx_;
  ctx.ctx_ = nullptr;
}

GridContext::~GridContext()
{
  if (ctx_)
    lb_grid_context_release(ctx_);
}

GridContext&
GridContext::operator=(const GridContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_)
    lb_grid_context_retain(ctx_);
  return *this;
}

GridContext
GridContext::create(const Config& config)
{
  auto* ptr = lb_grid_context_new(config);
  GridContext ctx(ptr);
  if (ptr)
    lb_grid_context_release(ptr);
  return ctx;
}

void
GridContext::submit(
  const SubmitGridOrderOptions& opts,
  AsyncCallback<GridContext, SubmitGridOrderResponse> callback) const
{
  lb_submit_grid_order_options_t opts2 = {
    opts.symbol.c_str(),
    opts.settlement_currency.c_str(),
    convert_grid_trade_rule(opts.grid_trading_rule),
  };

  lb_grid_context_submit(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, SubmitGridOrderResponse>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        SubmitGridOrderResponse resp =
          convert((const lb_submit_grid_order_response_t*)res->data);
        (*callback_ptr)(AsyncResult<GridContext, SubmitGridOrderResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<GridContext, SubmitGridOrderResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, SubmitGridOrderResponse>(callback));
}

void
GridContext::replace(
  const ReplaceGridOrderOptions& opts,
  AsyncCallback<GridContext, void> callback) const
{
  lb_replace_grid_order_options_t opts2 = {
    opts.order_id.c_str(),
    convert_grid_trade_rule(opts.grid_trading_rule),
  };

  lb_grid_context_replace(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, void>(res->userdata);
      (*callback_ptr)(AsyncResult<GridContext, void>(
        GridContext((const lb_grid_context_t*)res->ctx),
        Status(res->error),
        nullptr));
    },
    new AsyncCallback<GridContext, void>(callback));
}

void
GridContext::list(
  const std::optional<GetGridOrdersOptions>& opts,
  AsyncCallback<GridContext, GridOrdersResponse> callback) const
{
  lb_get_grid_orders_options_t opts2 = {
    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
  };
  lb_market_t market;

  if (opts) {
    opts2.page = opts->page ? &opts->page.value() : nullptr;
    opts2.limit = opts->limit ? &opts->limit.value() : nullptr;
    if (opts->market) {
      market = convert(*opts->market);
      opts2.market = &market;
    }
    opts2.status = opts->status ? opts->status->c_str() : nullptr;
    opts2.symbol = opts->symbol ? opts->symbol->c_str() : nullptr;
    opts2.sort_by = opts->sort_by ? opts->sort_by->c_str() : nullptr;
    opts2.sort_order = opts->sort_order ? opts->sort_order->c_str() : nullptr;
  }

  lb_grid_context_list(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, GridOrdersResponse>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        GridOrdersResponse resp =
          convert((const lb_grid_orders_response_t*)res->data);
        (*callback_ptr)(AsyncResult<GridContext, GridOrdersResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<GridContext, GridOrdersResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, GridOrdersResponse>(callback));
}

void
GridContext::list_by_ids(
  const GetGridOrdersByIdsOptions& opts,
  AsyncCallback<GridContext, std::vector<GridOrder>> callback) const
{
  std::vector<const char*> order_ids;
  std::transform(opts.order_ids.cbegin(),
                 opts.order_ids.cend(),
                 std::back_inserter(order_ids),
                 [](auto& id) { return id.c_str(); });
  lb_get_grid_orders_by_ids_options_t opts2 = {
    order_ids.data(),
    order_ids.size(),
  };

  lb_grid_context_list_by_ids(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, std::vector<GridOrder>>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto rows = (const lb_grid_order_t*)res->data;
        std::vector<GridOrder> rows2;
        std::transform(rows,
                       rows + res->length,
                       std::back_inserter(rows2),
                       [](auto& row) { return convert(&row); });

        (*callback_ptr)(AsyncResult<GridContext, std::vector<GridOrder>>(
          ctx, std::move(status), &rows2));
      } else {
        (*callback_ptr)(AsyncResult<GridContext, std::vector<GridOrder>>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, std::vector<GridOrder>>(callback));
}

void
GridContext::detail(
  const GetGridOrderDetailOptions& opts,
  AsyncCallback<GridContext, GridOrderDetail> callback) const
{
  lb_get_grid_order_detail_options_t opts2 = {
    opts.order_id.c_str(),
    opts.history_id ? opts.history_id->c_str() : nullptr,
    opts.limit ? &opts.limit.value() : nullptr,
  };

  lb_grid_context_detail(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, GridOrderDetail>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        GridOrderDetail resp =
          convert((const lb_grid_order_detail_t*)res->data);
        (*callback_ptr)(AsyncResult<GridContext, GridOrderDetail>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<GridContext, GridOrderDetail>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, GridOrderDetail>(callback));
}

void
GridContext::trigger_history(
  const GetGridTriggerHistoryOptions& opts,
  AsyncCallback<GridContext, GridTriggerHistoryResponse> callback) const
{
  lb_get_grid_trigger_history_options_t opts2 = {
    opts.grid_order_id.c_str(),
    opts.page ? &opts.page.value() : nullptr,
    opts.limit ? &opts.limit.value() : nullptr,
  };

  lb_grid_context_trigger_history(
    ctx_,
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext,
                                     GridTriggerHistoryResponse>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        GridTriggerHistoryResponse resp =
          convert((const lb_grid_trigger_history_response_t*)res->data);
        (*callback_ptr)(
          AsyncResult<GridContext, GridTriggerHistoryResponse>(
            ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(
          AsyncResult<GridContext, GridTriggerHistoryResponse>(
            ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, GridTriggerHistoryResponse>(callback));
}

void
GridContext::cancel(
  const std::string& order_id,
  AsyncCallback<GridContext, void> callback) const
{
  lb_grid_context_cancel(
    ctx_,
    order_id.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, void>(res->userdata);
      (*callback_ptr)(AsyncResult<GridContext, void>(
        GridContext((const lb_grid_context_t*)res->ctx),
        Status(res->error),
        nullptr));
    },
    new AsyncCallback<GridContext, void>(callback));
}

void
GridContext::suspend(
  const std::string& order_id,
  AsyncCallback<GridContext, void> callback) const
{
  lb_grid_context_suspend(
    ctx_,
    order_id.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, void>(res->userdata);
      (*callback_ptr)(AsyncResult<GridContext, void>(
        GridContext((const lb_grid_context_t*)res->ctx),
        Status(res->error),
        nullptr));
    },
    new AsyncCallback<GridContext, void>(callback));
}

void
GridContext::restart(
  const std::string& order_id,
  AsyncCallback<GridContext, void> callback) const
{
  lb_grid_context_restart(
    ctx_,
    order_id.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, void>(res->userdata);
      (*callback_ptr)(AsyncResult<GridContext, void>(
        GridContext((const lb_grid_context_t*)res->ctx),
        Status(res->error),
        nullptr));
    },
    new AsyncCallback<GridContext, void>(callback));
}

void
GridContext::symbol_info(
  const std::string& symbol,
  AsyncCallback<GridContext, GridSymbolInfo> callback) const
{
  lb_grid_context_symbol_info(
    ctx_,
    symbol.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<GridContext, GridSymbolInfo>(
          res->userdata);
      GridContext ctx((const lb_grid_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        GridSymbolInfo resp =
          convert((const lb_grid_symbol_info_t*)res->data);
        (*callback_ptr)(AsyncResult<GridContext, GridSymbolInfo>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<GridContext, GridSymbolInfo>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<GridContext, GridSymbolInfo>(callback));
}

} // namespace grid
} // namespace longbridge

File: java/src/grid_context.rs
use std::sync::Arc;

use jni::{
    JNIEnv,
    objects::{JClass, JObject, JString},
    sys::jobjectArray,
};
use longbridge::{
    Config, Market,
    grid::{
        GetGridOrderDetailOptions, GetGridOrdersByIdsOptions, GetGridOrdersOptions,
        GetGridTriggerHistoryOptions, GridContext, GridTradeRule, ReplaceGridOrderOptions,
        SubmitGridOrderOptions,
    },
};

use crate::{
    async_util,
    error::jni_result,
    types::{FromJValue, JavaInteger, JavaLong, ObjectArray, get_field},
};

struct ContextObj {
    ctx: GridContext,
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_newGridContext(
    mut env: JNIEnv,
    _class: JClass,
    config: i64,
) -> i64 {
    jni_result(&mut env, 0i64, |_env| {
        Ok(Box::into_raw(Box::new(ContextObj {
            ctx: GridContext::new(Arc::new((*(config as *const Config)).clone())),
        })) as i64)
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_freeGridContext(
    _env: JNIEnv,
    _class: JClass,
    ctx: i64,
) {
    let _ = Box::from_raw(ctx as *mut ContextObj);
}

/// Read a `GridTradeRule` from a Java `com.longbridge.grid.GridTradeRule`
/// object. Every field is optional (nullable box type on the Java side).
fn read_grid_trade_rule(
    env: &mut JNIEnv<'_>,
    obj: &JObject<'_>,
) -> jni::errors::Result<GridTradeRule> {
    fn read_opt_bool(
        env: &mut JNIEnv<'_>,
        obj: &JObject<'_>,
        name: &str,
    ) -> jni::errors::Result<Option<bool>> {
        let field = env.get_field(obj, name, "Ljava/lang/Boolean;")?.l()?;
        if field.is_null() {
            Ok(None)
        } else {
            Ok(Some(
                env.call_method(&field, "booleanValue", "()Z", &[])?.z()?,
            ))
        }
    }

    let mut rule = GridTradeRule::default();
    rule.submitted_base_price = get_field(env, obj, "submittedBasePrice")?;
    rule.upper_limit_price = get_field(env, obj, "upperLimitPrice")?;
    rule.lower_limit_price = get_field(env, obj, "lowerLimitPrice")?;
    rule.trigger_price_type = get_field::<_, _, Option<longbridge::grid::TriggerPriceType>>(
        env,
        obj,
        "triggerPriceType",
    )?;
    rule.trigger_spread_up = get_field(env, obj, "triggerSpreadUp")?;
    rule.trigger_spread_down = get_field(env, obj, "triggerSpreadDown")?;
    rule.trigger_percent_up = get_field(env, obj, "triggerPercentUp")?;
    rule.trigger_percent_down = get_field(env, obj, "triggerPercentDown")?;
    rule.multiple_trigger = read_opt_bool(env, obj, "multipleTrigger")?;
    rule.time_in_force =
        get_field::<_, _, Option<longbridge::grid::GridTimeInForce>>(env, obj, "timeInForce")?;
    rule.upper_limit_quantity = get_field(env, obj, "upperLimitQuantity")?;
    rule.lower_limit_quantity = get_field(env, obj, "lowerLimitQuantity")?;
    rule.expire_time = get_field::<_, _, Option<JavaLong>>(env, obj, "expireTime")?.map(i64::from);
    rule.upper_limit_event =
        get_field::<_, _, Option<longbridge::grid::GridLimitEvent>>(env, obj, "upperLimitEvent")?;
    rule.lower_limit_event =
        get_field::<_, _, Option<longbridge::grid::GridLimitEvent>>(env, obj, "lowerLimitEvent")?;
    rule.trigger_sell_depth =
        get_field::<_, _, Option<JavaInteger>>(env, obj, "triggerSellDepth")?.map(i32::from);
    rule.trigger_buy_depth =
        get_field::<_, _, Option<JavaInteger>>(env, obj, "triggerBuyDepth")?.map(i32::from);
    rule.trigger_quantity = get_field(env, obj, "triggerQuantity")?;
    rule.support_shortsell = read_opt_bool(env, obj, "supportShortsell")?;
    rule.rth = get_field::<_, _, Option<JavaInteger>>(env, obj, "rth")?.map(i32::from);
    rule.grid_order_type_up = get_field(env, obj, "gridOrderTypeUp")?;
    rule.grid_order_type_down = get_field(env, obj, "gridOrderTypeDown")?;
    Ok(rule)
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextSubmit(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let symbol: String = get_field(env, &opts, "symbol")?;
        let settlement_currency: String = get_field(env, &opts, "settlementCurrency")?;
        let rule_obj = env
            .get_field(
                &opts,
                "gridTradingRule",
                "Lcom/longbridge/grid/GridTradeRule;",
            )?
            .l()?;
        let rule = read_grid_trade_rule(env, &rule_obj)?;
        let new_opts = SubmitGridOrderOptions::new(symbol, settlement_currency, rule);
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.submit(new_opts).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextReplace(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_id: String = get_field(env, &opts, "orderId")?;
        let rule_obj = env
            .get_field(
                &opts,
                "gridTradingRule",
                "Lcom/longbridge/grid/GridTradeRule;",
            )?
            .l()?;
        let rule = read_grid_trade_rule(env, &rule_obj)?;
        let new_opts = ReplaceGridOrderOptions::new(order_id, rule);
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.replace(new_opts).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextList(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let opts = if !opts.is_null() {
            let mut new_opts = GetGridOrdersOptions::new();
            if let Some(page) = get_field::<_, _, Option<JavaInteger>>(env, &opts, "page")? {
                new_opts = new_opts.page(page.into());
            }
            if let Some(limit) = get_field::<_, _, Option<JavaInteger>>(env, &opts, "limit")? {
                new_opts = new_opts.limit(limit.into());
            }
            if let Some(market) = get_field::<_, _, Option<Market>>(env, &opts, "market")? {
                new_opts = new_opts.market(market);
            }
            if let Some(status) = get_field::<_, _, Option<String>>(env, &opts, "status")? {
                new_opts = new_opts.status(status);
            }
            if let Some(symbol) = get_field::<_, _, Option<String>>(env, &opts, "symbol")? {
                new_opts = new_opts.symbol(symbol);
            }
            if let Some(sort_by) = get_field::<_, _, Option<String>>(env, &opts, "sortBy")? {
                new_opts = new_opts.sort_by(sort_by);
            }
            if let Some(sort_order) = get_field::<_, _, Option<String>>(env, &opts, "sortOrder")? {
                new_opts = new_opts.sort_order(sort_order);
            }
            Some(new_opts)
        } else {
            None
        };
        async_util::execute(env, callback, async move {
            let resp = __owned_ctx.list(opts).await?;
            Ok(crate::types::GridOrdersResponse::new(
                resp.grid_order,
                resp.has_more,
            ))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextListByIds(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    order_ids: jobjectArray,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_ids: ObjectArray<String> =
            FromJValue::from_jvalue(env, JObject::from_raw(order_ids).into())?;
        let new_opts = GetGridOrdersByIdsOptions::new(order_ids.0);
        async_util::execute(env, callback, async move {
            Ok(ObjectArray(__owned_ctx.list_by_ids(new_opts).await?))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextDetail(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_id: String = get_field(env, &opts, "orderId")?;
        let mut new_opts = GetGridOrderDetailOptions::new(order_id);
        if let Some(history_id) = get_field::<_, _, Option<String>>(env, &opts, "historyId")? {
            new_opts = new_opts.history_id(history_id);
        }
        if let Some(limit) = get_field::<_, _, Option<JavaInteger>>(env, &opts, "limit")? {
            new_opts = new_opts.limit(limit.into());
        }
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.detail(new_opts).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextTriggerHistory(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let grid_order_id: String = get_field(env, &opts, "gridOrderId")?;
        let mut new_opts = GetGridTriggerHistoryOptions::new(grid_order_id);
        if let Some(page) = get_field::<_, _, Option<JavaInteger>>(env, &opts, "page")? {
            new_opts = new_opts.page(page.into());
        }
        if let Some(limit) = get_field::<_, _, Option<JavaInteger>>(env, &opts, "limit")? {
            new_opts = new_opts.limit(limit.into());
        }
        async_util::execute(env, callback, async move {
            let resp = __owned_ctx.trigger_history(new_opts).await?;
            Ok(crate::types::GridTriggerHistoryResponse::new(
                resp.trigger_orders,
                resp.has_more,
            ))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextCancel(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    order_id: JString,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_id: String = FromJValue::from_jvalue(env, order_id.into())?;
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.cancel(order_id).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextSuspend(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    order_id: JString,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_id: String = FromJValue::from_jvalue(env, order_id.into())?;
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.suspend(order_id).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextRestart(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    order_id: JString,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let order_id: String = FromJValue::from_jvalue(env, order_id.into())?;
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.restart(order_id).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_gridContextSymbolInfo(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    symbol: JString,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let __owned_ctx = context.ctx.clone();
        let symbol: String = FromJValue::from_jvalue(env, symbol.into())?;
        async_util::execute(env, callback, async move {
            Ok(__owned_ctx.symbol_info(symbol).await?)
        })?;
        Ok(())
    })
}
